backend-rust/src/indexer.rs

```diff
@@ -3501,81 +3638,155 @@ impl PreparedUpdateAccountBalance {
+                SET
+                    self_suspended = NULL,
+                    inactive_suspended = NULL,
+                    primed_for_suspension = $2
+                WHERE id=$1",
+            self.baker_id,
+            self.block_height
         )
         .execute(tx.as_mut())
         .await?;
-        self.account_statement.save(tx, transaction_index).await?;
         Ok(())
     }
 }
 
-/// Represent the event of a transfer of CCD from one account to another.
-pub struct PreparedCcdTransferEvent {
-    /// Updating the sender account balance.
-    update_sender:   PreparedUpdateAccountBalance,
-    /// Updating the receivers account balance.
-    update_receiver: PreparedUpdateAccountBalance,
+/// Represent the potential event of baker being "unprimed" for suspension.
+/// The baker of the block, plus the signers of the quarum certificate when
+/// included in the block. This might include baker IDs which are not primed at
+/// the time.
+struct PreparedUnmarkPrimedForSuspension {
+    baker_ids: Vec<i64>,
 }
 
-impl PreparedCcdTransferEvent {
-    pub fn prepare(
-        sender_address: Arc<String>,
-        receiver_address: Arc<String>,
-        amount: Amount,
-        block_height: AbsoluteBlockHeight,
-    ) -> anyhow::Result<Self> {
-        let amount: i64 = amount.micro_ccd().try_into()?;
-        let update_sender = PreparedUpdateAccountBalance::prepare(
-            sender_address.clone(),
-            -amount,
-            block_height,
-            AccountStatementEntryType::TransferOut,
-        )?;
-        let update_receiver = PreparedUpdateAccountBalance::prepare(
-            receiver_address.clone(),
-            amount,
-            block_height,
-            AccountStatementEntryType::TransferIn,
-        )?;
+impl PreparedUnmarkPrimedForSuspension {
+    fn prepare(data: &BlockData) -> anyhow::Result<Self> {
+        let mut baker_ids = Vec::new();
+        if let Some(baker_id) = data.block_info.block_baker {
+            baker_ids.push(baker_id.id.index.try_into()?);
+        }
+        if let Some(qc) = data.certificates.quorum_certificate.as_ref() {
+            for signer in qc.signatories.iter() {
+                baker_ids.push(signer.id.index.try_into()?);
+            }
+        }
         Ok(Self {
-            update_sender,
-            update_receiver,
+            baker_ids,
         })
     }
 
-    pub async fn save(
+    async fn save(
         &self,
         tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
-        transaction_index: i64,
     ) -> anyhow::Result<()> {
-        self.update_sender.save(tx, Some(transaction_index)).await?;
-        self.update_receiver.save(tx, Some(transaction_index)).await?;
+        sqlx::query!(
+            "UPDATE bakers
+                SET primed_for_suspension = NULL
+                WHERE
+                    primed_for_suspension IS NOT NULL
+                    AND id = ANY ($1)",
+            &self.baker_ids,
+        )
```

Comment:
It might be worth caching (some of the information from) the bakers table, so that there's no need to make an update query at all in the common case where nothing is being updated.

Reply:
Yes, you are probably right.
To avoid too much complexity, I'll add a partial index on the bakers where primed_for_suspension is not null.